### ISSUE TITLE
Is it possible to access Audio buffers to retrieve info for a VUMeter?

### ISSUE THREAD

**Issue by ruirodrigues-eyenov** [ThibaultBee/StreamPack#39]
Basically I would like to show a VUmeter with the audio volume/gain/amplitude being sent when streaming.
I could not found any way, I wonder if that is possible?
Thanks

**Comment by ThibaultBee** [comment]
Hello :)

I was kind of expecting this request.
There is a branch where you can get peak and/or rms value of each audio channels: https://github.com/ThibaultBee/StreamPack/tree/feature/vu-meter. I just rebased it.
The only things missing for now is that when preview is running, you won't get any value cause audio capture is not running and I still don't really know how to implement this part.

Regards,
Thibault

**Comment by ruirodrigues-eyenov** [comment]
Hello :)

this is great! exactly what I wanted! thanks :)

I don't need the values while previewing but from what I can understand you can just `startRecording` on the `AudioRecord` and discard the Frames after getting the measurements.

**Comment by ThibaultBee** [comment]
Hi,
This is the issue: there aren't any audio frames when encoder is not running. It means that you can only access audio frame when StreamPack is live streaming.

**Comment by ThibaultBee** [comment]
What metrics do you have in mind?

**Comment by ruirodrigues-eyenov** [comment]
latency, bandwidth, frame rate, bitrate. I am not sure if those are possible at all.

**Comment by ThibaultBee** [comment]
Do you use SRT or RTMP?
I guess it is quite easy to return latency, bandwidth and network bitrate with SRT, not sure about RTMP.

**Comment by ruirodrigues-eyenov** [comment]
SRT 😄

**Comment by ThibaultBee** [comment]
Starting from v3.1, an `AudioProcessor` will appear and it will give access to audio buffer.

**Comment by dimadesu** [comment]
@ThibaultBee Finally, I had time for this. I was able to get it to work. See this PR https://github.com/dimadesu/LifeStreamer/pull/272

Had to make a small adjustment on top your changes https://github.com/dimadesu/StreamPack/commit/9671a9de47350604aa20301b3326cc020c110dde

I don't know about this as much as you do, so not sure if everything is like proper and correct, but it works.

**Comment by ThibaultBee** [comment]
I have made some test on my side.

And found some issues which need to be tackle.

-  `java.lang.IllegalStateException: buffer is inaccessible`
You definitely need a deep copy of the bytebuffer before dispatching to the consomer effect scope.
Because in callback mode, the rawBuffer belongs to the codec input buffer pool and we don't want to hold for other purpose than encoding.
In pushpull mode, the buffer belongs to the audio buffer pool. With the current implementation, it could be put back in the pool despite the consumer effect is still using it. We could make a hook for that to keep the number of reference of the buffer and only release the buffer when all the users close it. But once again, it is not possible in callback mode.

- the access to the effect list must be thread safe (synchronize) to avoid a crash when accessing the list in the `process` while a effect is being removed from the list itself.

- Start the audio source to make the effects runs when the pipeline is not running.
In callback mode, the encoder schedules the calls to the audio source. Because of that, it won't be possible to run the effects without the while pipeline running in callback mode. The offline audio source will only be possible to run in pushpull mode.

**Comment by dimadesu** [comment]
At the moment, I run it only when streaming if that's what you're talking about.
It could be nice to run meters before stream start, but no big deal.
I think I am okay with this so far, my users didn't complain about it either.

**Comment by ThibaultBee** [comment]
2/ Has been also done.

The 3/ is the most difficult part. Let's say we postpone it for this iteration.

TODO list for later:

## Next iteration: VU meter run even when pipeline is not running.
 
- [ ] make audio input/source reconfigurable if it is already streaming. Beware the `setAudioSourceConfig` must not be called if an output is already using audio.
- [x] add an API to start/stop the audio input explicitly. Beware, a stopStream from the pipeline must not stop the audio input if the audio input `startStream` was explicitly called. (or add an API).

## Following iteration
- [ ] Pass the (re-)configuration to the effect

**Comment by ThibaultBee** [comment]
In the dev_v3_2 branch, I added a `startCapture` and a `stopCapture` on the `AudioInput` to make the audio source running when the pipeline is NOT streaming.
With that you can display a VUmeter when the pipeline is not streaming.

I will close this issue after v3.2 as the initial request has been adressed.

**Comment by dimadesu** [comment]
Great addition.
Tons of changes on that branch.

**Comment by ThibaultBee** [comment]
> Great addition. Tons of changes on that branch.

Indeed. Have you tested it? (stream and all)? Let's talk there: https://github.com/ThibaultBee/StreamPack/discussions/273

**Comment by dimadesu** [comment]
I haven't tested it yet. I was very busy recently. I want to play with it when I have more time.

**Comment by ThibaultBee** [comment]
Released in 3.2.0
